diesel/src/pg/expression/functions.rs

```diff
@@ -687,3 +688,28 @@ define_sql_function! {
     #[cfg(feature = "postgres_backend")]
     fn daterange(lower: Nullable<Date>, upper: Nullable<Date>, bound: RangeBoundEnum) -> Daterange;
 }
+
+define_sql_function! {
+    /// Append an element to the end of an array.
+    /// # Example
+    ///
+    /// ```rust
+    /// # include!("../../doctest_setup.rs");
+    /// #
+    /// # fn main() {
+    /// #     run_test().unwrap();
+    /// # }
+    /// #
+    /// # fn run_test() -> QueryResult<()> {
+    /// #     use diesel::dsl::array_append;
+    /// #     use diesel::sql_types::{Integer, Array};
+    /// #     let connection = &mut establish_connection();
+    /// let ints = diesel::select(array_append::<Integer, Array<_>, _, _>(vec![Some(1), Some(2)], Some(3)))
+    ///     .get_result::<Vec<Option<i32>>>(connection)?;
+    /// assert_eq!(vec![Some(1), Some(2), Some(3)], ints);
+    /// #     Ok(())
+    /// # }
+    /// ```
+    #[cfg(feature = "postgres_backend")]
+    fn array_append<T: SingleValue, Arr: ArrayOrNullableArray<Inner=Nullable<T>> + SingleValue>(a: Arr, e: Nullable<T>) -> Array<Nullable<T>>;
```

Comment:
It doesn't look like there should be Nullable here : the fact that pg arrays are always Nullable is already expressed to the Rust type system by the fact that we export them to the `schema.rs` as `Array<Nullable<T>>`, and if the schema was patched to express that the type of the array is in fact `Array<T>` we should still be able to use this function by providing an expression of sql type T.

Reply:
It looks like this can be achieved by having `Nullable<T>` as ElementType for ArrayOrNullableArray